README.md
# FastDub

Package for voice over subtitles:

* with the ability to embed in video,
* audio ducking,
* dynamic voice changer for a single track. _Add **--voice-set-anchor** at the beginning of the subtitle line. (Applies
  to
  all
  subsequent ones)_

> pip install -U [FastDub](https://pypi.org/project/FastDub/)

---

# Install for Ubuntu

> sudo apt update -y && sudo apt upgrade -y
>
> sudo apt install python3 python3-pip ffmpeg [espeak](http://espeak.sourceforge.net/data/) libxml2-dev libxslt1-dev
>
> ; libxml for translator functions
>
> sudo python3 -m pip install -U pip setuptools wheel
>
> sudo python3 -m pip install -U [FastDub](https://pypi.org/project/FastDub/)

# Install all dependencies

> pip install FastDub[ALL]  # default

# Youtube support

## Youtube argument group

> pip install FastDub[YT]

### Process all videos from a channel

_To get the channel id if it only has a username, run the JS code in the browser console:_

```javascript
document.querySelector("body>link").href
```

**Then, in the input parameter,
you can pass a link to the channel (with an id, not with a username) or a string of the format "c/CHANNEL_ID"**

_For example:_

> python -m fastdub -i "c/[UCi-5OZ2tYuwMLIcEyOsbdRA](https://www.youtube.com/channel/UCi-5OZ2tYuwMLIcEyOsbdRA)" -yt

## YouTube Search argument group

That the input data was a request to YouTube, they must begin with "?".

_For Example:_

> python -m fastdub -i "?#python" -yt -yts-l 5

## YouTube Upload argument group

> pip install FastDub[YTU]

To download, you need to go to [credentials](https://console.cloud.google.com/apis/credentials) (Create a new project if
needed) > <kbd>+ Create credentials</kbd> > <kbd>OAuth Client ID</kbd> > <kbd>Desktop App</kbd>

After filling in the required fields, <kbd>DOWNLOAD JSON</kbd> format and save to a working folder under the name
_client_secrets.json_

* Specifying the -ytu-ps argument as non-private may result in processing errors, but this was not observed during
  testing.

* Please note that video uploads require a large amount of [quota](https://console.cloud.google.com/iam-admin/quotas) (
  default has 10,000 per day)

# Subtitles translate

### Translate argument group

> pip install FastDub[TR]

# Usage

> python -m fastdub --help

```
usage: fastdub [-h] [-rc {0,1,2}] [-ra | --cleanup-audio | -n-ra | --no-cleanup-audio] [-ev | --export-video | -n-ev | --no-export-video]
               [-l LANGUAGE] [-ll {CRITICAL,FATAL,ERROR,WARN,WARNING,INFO,DEBUG,NOTSET}] [-tc THREADS_COUNT] -i INPUT [-vf VIDEO_FORMAT]
               [-sf SUBTITLES_FORMAT] [-En EXCLUDE [EXCLUDE ...]] [-Eu EXCLUDE_UNDERSCORE] [-sc | --sidechain | -n-sc | --no-sidechain]
               [-sc-args SIDECHAIN_FFMPEG_PARAMS] [-sc-lvl SIDECHAIN_LEVEL_SC]
               [-v {...}]
               [-a ALIGN] [-v-set-a VOICE_SET_ANCHOR] [-fll {trace,debug,verbose,info,warning,error,fatal,panic,quiet}]
               [-y | --confirm | -n-y | --no-confirm] [-af AUDIO_FORMAT] [-wm WATERMARK] [-tb | --traceback | -n-tb | --no-traceback] [-yt]      
               [-ak API_KEYS [API_KEYS ...]] [-yts] [-yts-l YOUTUBE_SEARCH_LIMIT] [-yts-rg YOUTUBE_SEARCH_REGION] [-ytu]
               [-ytu-ps {private,public,unlisted}] [-ytu-t] [-tr] [--rewrite-srt | --no-rewrite-srt]
               [-ts {...}]

fastdub is a tool for dubbing videos by subtitle files.

optional arguments:
  -h, --help            show this help message and exit
  -rc {0,1,2}, --remove-cache {0,1,2}
                        Remove all cache (_cached_texts directory) files
                                0 - No remove cache
                                1 - Delete cache before voice acting
                                2 - Delete cache after voice acting (default)
  -ra, --cleanup-audio, -n-ra, --no-cleanup-audio
                        Remove result audio if video exists (default True) (default: True)
  -ev, --export-video, -n-ev, --no-export-video
  -l LANGUAGE, --language LANGUAGE
                        Subtitles language (ru)
  -ll {CRITICAL,FATAL,ERROR,WARN,WARNING,INFO,DEBUG,NOTSET}, --loglevel {CRITICAL,FATAL,ERROR,WARN,WARNING,INFO,DEBUG,NOTSET}
                        Program log level
  -tc THREADS_COUNT, --threads-count THREADS_COUNT
                        Process count to download (pass to cpu count, < 2 to disable)
                                *N = N * cpu count

Input:
  -i INPUT, --input INPUT
                        Input directory/YouTube Playlist/Video URL.
  -vf VIDEO_FORMAT, --video-format VIDEO_FORMAT
                        Video format (default: .mp4).
  -sf SUBTITLES_FORMAT, --subtitles-format SUBTITLES_FORMAT
                        Subtitles format (default: .srt).

Input Exclude:
  -En EXCLUDE [EXCLUDE ...], --exclude EXCLUDE [EXCLUDE ...]
                        Exclude <name> (glob)
  -Eu EXCLUDE_UNDERSCORE, --exclude-underscore EXCLUDE_UNDERSCORE
                        Exclude files starts with underscore

Audio Ducking:
  -sc, --sidechain, -n-sc, --no-sidechain
                        Enable audio side chain compress (ducking) (default: True)
  -sc-args SIDECHAIN_FFMPEG_PARAMS, --sidechain-ffmpeg-params SIDECHAIN_FFMPEG_PARAMS
                        sidechain FFmpeg parameters (default '')
  -sc-lvl SIDECHAIN_LEVEL_SC, --sidechain-level-sc SIDECHAIN_LEVEL_SC
                        Set sidechain gain. Range is between 0.015625 and 64. (default 0.8)

Voicer:
  -v {...}, --voice {...}
                        SAPI voice for voice acting.
  -a ALIGN, --align ALIGN
                        Audio fit align (divisor)
                                1 = right
                                2 = center (default)
  -v-set-a VOICE_SET_ANCHOR, --voice-set-anchor VOICE_SET_ANCHOR
                        Anchor indicating voice actor change (default "!:")

FFmpeg Output:
  -fll {trace,debug,verbose,info,warning,error,fatal,panic,quiet}, --ffmpeg-loglevel {trace,debug,verbose,info,warning,error,fatal,panic,quiet}  
                        FFmpegWrapper loglevel
  -y, --confirm, -n-y, --no-confirm
                        Don't ask for confirmation (default: True)
  -af AUDIO_FORMAT, --audio-format AUDIO_FORMAT
                        Out audio files format (default wav)
  -wm WATERMARK, --watermark WATERMARK
                        Add watermark to output video

Terminal Output:
  -tb, --traceback, -n-tb, --no-traceback
                        Show debug traceback (default: False)

YouTube:
  -yt, --youtube
  -ak API_KEYS [API_KEYS ...], --api-keys API_KEYS [API_KEYS ...]
                        Youtube API key/s

YouTube Search:
  -yts, --youtube-search
                        The input data is a query for searching on YouTube. (Adds "?" at the start of input)
  -yts-l YOUTUBE_SEARCH_LIMIT, --youtube-search-limit YOUTUBE_SEARCH_LIMIT
                        Sets limit to the number of results. Defaults to 20.
  -yts-rg YOUTUBE_SEARCH_REGION, --youtube-search-region YOUTUBE_SEARCH_REGION
                        Sets the result region. Defaults to "US".

YouTube Upload:
  -ytu, --youtube-upload
                        yt_upload video to YouTube channel after voice acting.
  -ytu-ps {private,public,unlisted}, --privacy-status {private,public,unlisted}
                        Video privacy status (If not private, errors are possible)
  -ytu-t, --youtube-upload-translate
                        Translate title and description on upload. (+ Arguments from translate argument group)

Translate:
  -tr, --translate      Translate input subtitles files
  --rewrite-srt, --no-rewrite-srt
                        Rewrite input subtitles files.
                        If not, add "_" to the beginning of the original subtitle file. (default: False)
  -ts {...}, --translate-service {...}
                        Subtitle translation service. (default google)
```

**If the voice set after "!:" is not selected during voiceover, clear the cache with the -rc argument**

## Example

> python -m fastdub -i DirToDub -v "[Yuriy](https://rhvoice.su/downloads/?voice=yuriy&type=sapi)" --no-sidechain -sf vtt
> -vf mkv

All files in the folder will be voices (*.mkv, *.vtt)

Default is mp4 video and srt subtitles.

Then:
_name and name2 will be announced, and the results will be saved to a folder with the appropriate name._

## YT Example

> python -m fastdub **-yt** -i https://youtu.be/DD5UKQggXTc
> -v "[Yuriy](https://rhvoice.su/downloads/?voice=yuriy&type=sapi)"
> -l ru

### YouTube Search Example

> python -m fastdub -yt **-yts** -i "#annoyingorange" -l ru

## Translate Example

> python -m fastdub -i DirToDub -tr **-ts iciba** _-l [ru](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes)_

Default service is Google.

## Other

### The `fastdub.utils` module stores helper functions.

### You can write your issues on [GitHub](https://github.com/NIKDISSV-Forever/fastdub/issues) in English or in Russian.


CHANGELOG.md
# 3.8.0

## General

- `changelog` has been translated into English.
- Added `logging.error`

## FFmpeg Wrapper

- `ffmpeg_wrapper` module objects renamed according to PEP-8
  (`FFMpeg` to `FFmpeg`)

## Translator

- `translator.get_service_by_name` is now a valid function even if required dependencies are not installed, in which
  case if call it will raise `KeyError`

## YouTube

- dependency `pafy` added to `extra_requires/requires_yt.txt`

# 3.7.0

## General

- **Changed minimum supported Python version from 3.8 to 3.9**

## CLI

- Fixed bug with `logging.getLevelNamesMapping()` in `__main__.py`

# 3.6.0

## General

- Package `logging` used
- Bugfix
- optimization

## Audio Processing

- `sidechain` is now much faster using `ffmpeg`

## FFmpeg Wrapper

- The `ffmpeg-imageio` library is no longer used, instead the `ffmpeg` command is always called

## CLI

- Argument `--sidechain-ffmpeg-params` see [FFmpeg Documentation](https://ffmpeg.org/ffmpeg-all.html#sidechaincompress)
- Separate `--sidechain-level-sc` see `level_sc` higher.
- `-ll` is now program loglevel and ffmpeg loglevel is now `-fll`

# 3.5.2

## Translator

- Now it will not cause an error in the absence of the Internet.

## Subtitles

- Fixed possible .srt format differences (now `.` is allowed instead of `,`)

# 3.5.1

## Subtitles

- Fixed possible problems with different encodings (added `chardet` module)

# 3.5.0

## Voiceover Process

- Added argument `--export-video` (`-ev`): `BooleanOptionalAction` if `False` (`--no-export-video`, `-n-ev`)
  only exports audio

# 3.4.2

- Minor improvements and optimizations.

## Subtitles

- The `TimeLabel` class is no longer a subclass of `Line`

# 3.4.1

- Fixed documentation

# 3.4.0

## General

- Added the ability to apply a watermark to the video (see the `--watermark` argument)

## CLI

- Bugfix
- The `--watermark` argument. The text of the watermark is passed to it.
- Now, when launched from the console, the FastDub mini-banner is shown.

## Voiceover Process

- Bugfix
- Now fitted audio (which starts with '_') is always deleted as it's a temporary file

## FFmpeg Wrapper

- Added class method `FFmpegWrapper.get_video_duration_s`
  which is used in `FFmpegWrapper.get_video_duration_ms`
- The video result includes: dubbed audio, original audio, subtitles, original subtitles (if any)

## Submodules

### Translator

- The latest version of the translators module is now used

# 3.3.0

## General

- Implemented class `GlobalSettings`

## Audio Processing

- The `calc_speed_change_ffmpeg_arg` function for calculating the `-af` ffmpeg argument
  for speed change has been changed.

- Significantly accelerated using a different algorithm.

- Optimized `side_chain` function.

## Voiceover Process

- Audio length trimming and TTS are now combined into one process to increase performance.
- Now a subfolder `_result_working_dir` is created in which separate already fitted subtitle lines are saved, after
  which all parts are glued together with ffmpeg.
  Due to this, memory consumption is significantly reduced.
- `moviepy` is no longer a dependency, ffmpeg output is parsed instead

## FFmpeg Wrapper

- Now all arguments are converted to strings before calling `check_call`

# 3.2.0

## CLI

- Omitted optional `type=str` arguments.
- Changed `BooleanOptionalAction` logic, both `--no-` and `-n-` negation are now available.
- The `--cleanup-audio` argument is now `BooleanOptionalAction` (default `True`), `--cleanup-level` has been removed.
- Argument `--voice-set-anchor` (also `!:` by default).
- `--audio-format` argument for output audio (default `wav`).
- The `--sidechain` argument has been renamed to `--side-chain`.
- Added shorthand to `--traceback` argument - `-tb`.

## Voiceover Process

- Bugfix.
- Optimization of track length fitting (Fitting).

## TTS

- `anchor` argument (see `--voice-set-anchor`).
- Voice change optimization.

## Submodules

### YouTube

- When multi-loading, fully loaded videos are removed from the screen (so as not to take up space in vain).


fastdub/__main__.py
from __future__ import annotations

import argparse
import logging
import sys
from glob import iglob
from multiprocessing import cpu_count
from os import getcwd
from time import perf_counter

import rich.traceback

import fastdub.youtube
from fastdub import GlobalSettings, PrettyViewPrefix, dubber, translator, voicer, youtube
from fastdub.ffmpeg_wrapper import DefaultFFmpegParams
from fastdub.translator.subs_translate import SrtTranslate

__all__ = ('parse_args', 'main')
THREADS_MAY_NEED = youtube.SUPPORTED or translator.SUPPORTED

if hasattr(logging, '_nameToLevel'):
    def _get_logging_level_names() -> tuple[str]:
        # noinspection PyProtectedMember
        return *logging._nameToLevel,
elif sys.version_info >= (3, 11):
    def _get_logging_level_names() -> tuple[str]:
        return *logging.getLevelNamesMapping(),
else:
    def _get_logging_level_names() -> tuple[str]:
        return 'INFO',


class BooleanOptionalAction(argparse.Action):
    __slots__ = ()

    def __init__(self, option_strings, dest, default=None, type=None,
                 choices=None, required=False, help=None, metavar=None):
        _option_strings = (*option_strings,
                           *('--no-' + option_string[2:] if option_string.startswith('--') else
                             '-n-' + option_string[1:] if option_string.startswith('-') else option_strings
                             for option_string in option_strings))
        if help is not None and default is not None and default is not argparse.SUPPRESS:
            help += " (default: %(default)s)"
        super().__init__(option_strings=_option_strings, dest=dest, nargs=0, default=default, type=type,
                         choices=choices, required=required, help=help, metavar=metavar)

    def __call__(self, parser, namespace, values, option_string=None):
        if option_string in self.option_strings:
            setattr(namespace, self.dest,
                    not (option_string.startswith('--no-') or option_string.startswith('-n-')))

    def format_usage(self):
        return ' | '.join(self.option_strings)


def _thread_count_type(tc: str) -> int:
    return int(cpu_count() * float(tc.removeprefix('*'))) if tc.startswith('*') else int(tc)


# noinspection PyTypeChecker
def parse_args() -> argparse.Namespace:
    arg_parser = argparse.ArgumentParser('fastdub',
                                         description='fastdub is a tool for dubbing videos by subtitle files.',
                                         formatter_class=argparse.RawTextHelpFormatter)

    arg_parser.add_argument('-rc', '--remove-cache', default=2, type=int, choices=(0, 1, 2),
                            help='Remove all cache (_cached_texts directory) files\n'
                                 '\t0 - No remove cache\n'
                                 '\t1 - Delete cache before voice acting\n'
                                 '\t2 - Delete cache after voice acting (default)')
    arg_parser.add_argument('-ra', '--cleanup-audio', action=BooleanOptionalAction, default=True,
                            help='Remove result audio if video exists (default True)')
    arg_parser.add_argument('-ev', '--export-video', action=BooleanOptionalAction, default=True)
    arg_parser.add_argument('-l', '--language', default='ru',
                            help='Subtitles language (ru)')
    arg_parser.add_argument('-ll', '--loglevel', default='INFO', type=str.upper,
                            choices=_get_logging_level_names(),
                            help='Program log level')

    if THREADS_MAY_NEED:
        arg_parser.add_argument('-tc', '--threads-count', default=cpu_count(),
                                type=_thread_count_type,
                                help='Process count to download (pass to cpu count, < 2 to disable)\n'
                                     '\t*N = N * cpu count')

    input_group = arg_parser.add_argument_group('Input')
    input_group.add_argument('-i', '--input', default=getcwd(), required=True,
                             help=f"Input directory{('/YouTube Playlist/Video URL' if youtube.SUPPORTED else '')}.")
    input_group.add_argument('-vf', '--video-format', default='.mp4',
                             help='Video format (default: .mp4).')
    input_group.add_argument('-sf', '--subtitles-format', default='.srt',
                             help='Subtitles format (default: .srt).')

    exclude_input_group = arg_parser.add_argument_group('Input Exclude')
    exclude_input_group.add_argument('-En', '--exclude', type=iglob, nargs='+', default=(),
                                     help='Exclude <name> (glob)')
    exclude_input_group.add_argument('-Eu', '--exclude-underscore', default=True,
                                     help='Exclude files starts with underscore')

    ducking_group = arg_parser.add_argument_group('Audio Ducking')
    ducking_group.add_argument('-sc', '--sidechain', action=BooleanOptionalAction, default=True,
                               help='Enable audio side chain compress (ducking)')
    ducking_group.add_argument('-sc-args', '--sidechain-ffmpeg-params', default=DefaultFFmpegParams.sidechain_args,
                               help=f'sidechain FFmpeg parameters (default {DefaultFFmpegParams.sidechain_args!r})')
    ducking_group.add_argument('-sc-lvl', '--sidechain-level-sc', type=float, default=.8,
                               help='Set sidechain gain. Range is between 0.015625 and 64. (default 0.8)')

    voicer_group = arg_parser.add_argument_group('Voicer')
    voicer_group.add_argument('-v', '--voice', type=str.lower, choices=voicer.VOICES_NAMES.keys(),
                              help='SAPI voice for voice acting.')
    voicer_group.add_argument('-a', '--align', default=2., type=float,
                              help='Audio fit align (divisor)'
                                   '\n\t1 = right'
                                   '\n\t2 = center (default)')
    voicer_group.add_argument('-v-set-a', '--voice-set-anchor', default='!:',
                              help='Anchor indicating voice actor change (default "!:")')

    ffmpeg_group = arg_parser.add_argument_group('FFmpeg Output')
    ffmpeg_group.add_argument('-fll', '--ffmpeg-loglevel', default='panic',
                              choices=(
                                  'trace', 'debug', 'verbose', 'info', 'warning', 'error', 'fatal', 'panic', 'quiet'),
                              help='FFmpegWrapper loglevel')
    ffmpeg_group.add_argument('-y', '--confirm', action=BooleanOptionalAction, default=True,
                              help="Don't ask for confirmation")
    ffmpeg_group.add_argument('-af', '--audio-format', default='wav',
                              help='Out audio files format (default wav)')
    ffmpeg_group.add_argument('-wm', '--watermark', default='',
                              help='Add watermark to output video')

    output_group = arg_parser.add_argument_group('Terminal Output')
    output_group.add_argument('-tb', '--traceback', action=BooleanOptionalAction, default=False,
                              help='Show debug traceback')

    if youtube.SUPPORTED:
        yt_group = arg_parser.add_argument_group('YouTube')
        yt_group.add_argument('-yt', '--youtube', action='store_true', default=False)
        yt_group.add_argument('-ak', '--api-keys', nargs='+', default=(), help='Youtube API key/s')

        yt_search_group = arg_parser.add_argument_group('YouTube Search')
        yt_search_group.add_argument('-yts', '--youtube-search', action='store_true', default=False,
                                     help='The input data is a query for searching on YouTube.'
                                          ' (Adds "?" at the start of input)')
        yt_search_group.add_argument('-yts-l', '--youtube-search-limit', type=int, default=20,
                                     help='Sets limit to the number of results. Defaults to 20.')
        yt_search_group.add_argument('-yts-rg', '--youtube-search-region', default='US',
                                     help='Sets the result region. Defaults to "US".')

    if fastdub.youtube.yt_upload.SUPPORTED:
        ytu_group = arg_parser.add_argument_group('YouTube Upload')
        ytu_group.add_argument('-ytu', '--youtube-upload', action='store_true', default=False,
                               help='yt_upload video to YouTube channel after voice acting.')
        ytu_group.add_argument('-ytu-ps', '--privacy-status',
                               default=fastdub.youtube.yt_upload.VALID_PRIVACY_STATUSES[0],
                               choices=fastdub.youtube.yt_upload.VALID_PRIVACY_STATUSES,
                               help='Video privacy status (If not private, errors are possible)')
        if translator.SUPPORTED:
            ytu_group.add_argument('-ytu-t', '--youtube-upload-translate', action='store_true', default=False,
                                   help='Translate title and description on upload. '
                                        '(+ Arguments from translate argument group)')

    if translator.SUPPORTED:
        translate_group = arg_parser.add_argument_group('Translate')
        translate_group.add_argument('-tr', '--translate', action='store_true', default=False,
                                     help='Translate input subtitles files')
        translate_group.add_argument('--rewrite-srt', action=BooleanOptionalAction, default=False,
                                     help='Rewrite input subtitles files.\n'
                                          'If not, add "_" to the beginning of the original subtitle file.')
        translate_group.add_argument('-ts', '--translate-service',
                                     default='google',
                                     choices=translator.SERVICES,
                                     help='Subtitle translation service. (default google)')
    return arg_parser.parse_args()


def banner():
    rich.print('[b][i]FastDub[/][/]\n'
               'https://github.com/NIKDISSV-Forever/FastDub')


def main():
    args = parse_args()

    if args.traceback:
        rich.traceback.install(show_locals=True)
    else:
        rich.traceback.install(extra_lines=0, word_wrap=True)
    logging.basicConfig(format='%(levelname)s: [%(asctime)s] %(message)s')
    logging.getLogger().setLevel(args.loglevel)

    remove_cache = args.remove_cache
    if remove_cache == 1:
        dubber.VOICER.cleanup()

    GlobalSettings.watermark = args.watermark
    DefaultFFmpegParams.ffmpeg_log_level = args.ffmpeg_loglevel

    total_time = 0
    if args.confirm:
        DefaultFFmpegParams.args += '-y',
        total_time = perf_counter()

    if THREADS_MAY_NEED:
        GlobalSettings.threads_count = args.threads_count

    if youtube.SUPPORTED and args.youtube:
        query: str = args.input
        if args.youtube_search and not query.startswith('?'):
            query = f'?{query.strip()}'
        youtube.downloader.DownloadYTVideo.API_KEYS |= {*args.api_keys}
        downloader = youtube.downloader.DownloadYTVideo(query,
                                                        args.language,
                                                        args.youtube_search_limit, args.youtube_search_limit, )
        downloader.multiprocessing_download()
        args.input = downloader.save_dir

    video_format = video_format if (video_format := args.video_format).startswith('.') else f'.{video_format}'
    subtitles_format = (
        subtitles_format if (subtitles_format := args.subtitles_format).startswith('.') else f'.{subtitles_format}')
    audio_format: str = args.audio_format.removeprefix('.')

    videos = dubber.Dubber.collect_videos(args.input, args.exclude_underscore, (*sum(args.exclude, []),))

    if translator.SUPPORTED:
        translate_service = translator.get_service_by_name(args.translate_service)
        if args.translate:
            SrtTranslate(args.language, translate_service, args.rewrite_srt).translate_dir(videos, subtitles_format)
    else:
        translate_service = None

    dubber.Dubber(args.voice, args.language, audio_format,
                  args.sidechain, args.sidechain_level_sc, args.sidechain_ffmpeg_params,
                  args.align,
                  args.cleanup_audio, args.export_video
                  ).dub_dir(videos, video_format, subtitles_format)

    if remove_cache == 2:
        voicer.Voicer().cleanup()

    if fastdub.youtube.yt_upload.SUPPORTED and args.youtube_upload:
        fastdub.youtube.yt_upload.uploader.Uploader(args.privacy_status,
                                                    translator.SUPPORTED and args.youtube_upload_translate,
                                                    translate_service
                                                    ).upload(args.input)

    if total_time:
        logging.info(f'total time {PrettyViewPrefix.from_seconds(perf_counter() - total_time)}')


if __name__ == '__main__':
    banner()
    main()
